# Data Wrangling : Clinical trial data for a new oral insulin

## Introduction

The increasing prevalence of diabetes in the 21st century is a problem, an epidemic even. Pre 1920s, diabetes was a feared disease, that most certainly led to death. Luckily, in the 1920s, a secretion in the pancreas that lowered blood sugar levels, soon to be called insulin, was discovered by soon to be Nobel Prize winner Frederick Banting.

People with severe diabetes, and only days left to live were saved, but the default method of administration back then was a needle, multiple times a day, and it still is now. This is scary for some people and uncomfortable and inconvenient for the vast majority. The future : Oral insulin. This is an active area of research, and has been for a long time. Historically though there's been a big roadblock, getting insulin through the stomachs thick lining.

In this project, we've got our hands on the phase two clinical trial data for a new innovative oral insulin called auralin. Auralin researchers believe their proprietary capsule will solve this stomach lining problem. Phase two trials test the efficacy and the dose response of a drug, plus identify common short term side effects, also known as adverse reactions. These typically involve several hundred patients.

350 patients participated in this clinical trial. None of the patients were using Novodra (a popular injectable insulin) or Auralin (the oral insulin being researched) as their primary source of insulin before. All were experiencing elevated HbA1c levels.

All 350 patients were treated with Novodra to establish a baseline HbA1c level and insulin dose. After four weeks, which isn’t enough time to capture all the change in HbA1c that can be attributed by the switch to Auralin or Novodra:

175 patients switched to Auralin for 24 weeks
175 patients continued using Novodra for 24 weeks

## Part I - Gathering data

This Auralin Phase II clinical trial dataset comes in three tables: `patients`, `treatments`, and `adverse_reactions`. Acquaint yourself with them through visual assessment below.

In [1]:
import pandas as pd

In [2]:
patients = pd.read_csv('patients.csv')
treatments = pd.read_csv('treatments.csv')
adverse_reactions = pd.read_csv('adverse_reactions.csv')

In [3]:
patients.head()

patient_id assigned_sex given_name  surname                    address  \
0           1       female        Zoe  Wellish       576 Brown Bear Drive   
1           2       female     Pamela     Hill  2370 University Hill Road   
2           3         male        Jae   Debord      1493 Poling Farm Road   
3           4         male       Liêm     Phan        2335 Webster Street   
4           5         male        Tim  Neudorf       1428 Turkey Pen Lane   

                city       state  zip_code        country  \
0  Rancho California  California   92390.0  United States   
1          Armstrong    Illinois   61812.0  United States   
2               York    Nebraska   68467.0  United States   
3         Woodbridge          NJ    7095.0  United States   
4             Dothan          AL   36303.0  United States   

                                      contact  birthdate  weight  height   bmi  
0        951-719-9170ZoeWellish@superrito.com  7/10/1976   121.7      66  19.6  
1       PamelaSHill@cuvox.de+1 (217) 569-3204   4/3/1967   118.8      66  19.2  
2            402-363-6804JaeMDebord@gustr.com  2/19/1980   177.8      71  24.8  
3  PhanBaLiem@jourrapide.com+1 (732) 636-8246  7/26/1951   220.9      70  31.7  
4             334-515-7487TimNeudorf@cuvox.de  2/18/1928   192.3      27  26.1

In [4]:
patients.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 14 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   patient_id    503 non-null    int64  
 1   assigned_sex  503 non-null    object 
 2   given_name    503 non-null    object 
 3   surname       503 non-null    object 
 4   address       491 non-null    object 
 5   city          491 non-null    object 
 6   state         491 non-null    object 
 7   zip_code      491 non-null    float64
 8   country       491 non-null    object 
 9   contact       491 non-null    object 
 10  birthdate     503 non-null    object 
 11  weight        503 non-null    float64
 12  height        503 non-null    int64  
 13  bmi           503 non-null    float64
dtypes: float64(3), int64(2), object(9)
memory usage: 55.1+ KB


- Missing demographic info
- Erroneous datatype (assigned_sex, zipcode, birthdate)

In [5]:
patients.describe()

patient_id      zip_code      weight      height         bmi
count  503.000000    491.000000  503.000000  503.000000  503.000000
mean   252.000000  49084.118126  173.434990   66.634195   27.483897
std    145.347859  30265.807442   33.916741    4.411297    5.276438
min      1.000000   1002.000000   48.800000   27.000000   17.100000
25%    126.500000  21920.500000  149.300000   63.000000   23.300000
50%    252.000000  48057.000000  175.300000   67.000000   27.200000
75%    377.500000  75679.000000  199.500000   70.000000   31.750000
max    503.000000  99701.000000  255.900000   79.000000   37.700000


- zip code is a float instead of string
- zip code sometimes is 4 digits instead of 5 ( minimum value 1002)
- height variable minimum value is 27 inches !!! given that the inclusion criteria of this clinical trial is 18 years old this doesn't seems like a plausible adult height.
- weight variable minimum value is 48.8 lbs !!! given that the inclusion criteria of this clinical trial is 18 years old this doesn't seems like a plausible adult weight.

In [6]:
patients.query('height == 27')

patient_id assigned_sex given_name  surname               address    city  \
4           5         male        Tim  Neudorf  1428 Turkey Pen Lane  Dothan   

  state  zip_code        country                          contact  birthdate  \
4    AL   36303.0  United States  334-515-7487TimNeudorf@cuvox.de  2/18/1928   

   weight  height   bmi  
4   192.3      27  26.1

In [7]:

weight_lbs = 192.3
height_in = 27
BMI = 703 * weight_lbs / (height_in*height_in)
BMI

185.44156378600823

In [8]:
# A BMI of 185 it's just can't be possible. Maybe it's a data entry error : 27 instead of 72

height_in = 72
BMI_check = 703 * weight_lbs / (height_in*height_in)
BMI_check


26.077719907407406

In [9]:
patients.query("weight == 48.8")

patient_id assigned_sex given_name   surname              address  \
210         211       female    Camilla  Zaitseva  4689 Briarhill Lane   

        city state  zip_code        country  \
210  Wooster    OH   44691.0  United States   

                                       contact   birthdate  weight  height  \
210  330-202-2145CamillaZaitseva@superrito.com  11/26/1938    48.8      63   

      bmi  
210  19.1

In [10]:
weight_lbs = 48.8*2.20462 # 2.20462 is the concersion factor between kilograms and pounds
height_in = 63
BMI = 703 * weight_lbs / (height_in*height_in)
BMI

19.05582654774502

In [11]:

patients.query("weight == 48.8").bmi


210    19.1
Name: bmi, dtype: float64

In [12]:
patients.state.value_counts()

state
California    36
TX            32
New York      25
CA            24
NY            22
MA            22
PA            18
GA            15
OH            14
Illinois      14
LA            13
MI            13
OK            13
Florida       13
NJ            12
VA            11
IL            10
MS            10
WI            10
AL             9
TN             9
IN             9
MN             9
FL             9
KY             8
NC             8
WA             8
MO             7
KS             6
NV             6
ID             6
CT             5
SC             5
IA             5
AR             4
ME             4
Nebraska       4
AZ             4
CO             4
ND             4
RI             4
WV             3
OR             3
MD             3
SD             3
DE             3
VT             2
MT             2
DC             2
NE             2
NM             1
WY             1
AK             1
NH             1
Name: count, dtype: int64


there are two representation name for some states : the full name and the abreviation. these states are -California : CA -New York : NY -Illinois : IL -Florida : FL -Nebraska : NE

In [13]:
patients[patients.address.duplicated()]

patient_id assigned_sex given_name     surname             address  \
29           30         male       Jake    Jakobsen   648 Old Dear Lane   
219         220         male         Mỹ       Quynh                 NaN   
229         230         male       John         Doe     123 Main Street   
230         231       female  Elisabeth     Knudsen                 NaN   
234         235       female    Martina   Tománková                 NaN   
237         238         male       John         Doe     123 Main Street   
242         243         male       John     O'Brian                 NaN   
244         245         male       John         Doe     123 Main Street   
249         250         male   Benjamin      Mehler                 NaN   
251         252         male       John         Doe     123 Main Street   
257         258         male        Jin        Kung                 NaN   
264         265       female   Wafiyyah      Asfour                 NaN   
269         270       female     Flavia  Fiorentino                 NaN   
277         278         male       John         Doe     123 Main Street   
278         279       female   Generosa       Cabán                 NaN   
282         283       female      Sandy      Taylor  2476 Fulton Street   
286         287         male      Lewis        Webb                 NaN   
296         297       female        Chỉ         Lâm                 NaN   
502         503         male        Pat     Gersten   2778 North Avenue   

            city     state  zip_code        country  \
29   Port Jervis  New York   12771.0  United States   
219          NaN       NaN       NaN            NaN   
229     New York        NY   12345.0  United States   
230          NaN       NaN       NaN            NaN   
234          NaN       NaN       NaN            NaN   
237     New York        NY   12345.0  United States   
242          NaN       NaN       NaN            NaN   
244     New York        NY   12345.0  United States   
249          NaN       NaN       NaN            NaN   
251     New York        NY   12345.0  United States   
257          NaN       NaN       NaN            NaN   
264          NaN       NaN       NaN            NaN   
269          NaN       NaN       NaN            NaN   
277     New York        NY   12345.0  United States   
278          NaN       NaN       NaN            NaN   
282     Rainelle        WV   25962.0  United States   
286          NaN       NaN       NaN            NaN   
296          NaN       NaN       NaN            NaN   
502         Burr  Nebraska   68324.0  United States   

                                        contact   birthdate  weight  height  \
29   JakobCJakobsen@einrot.com+1 (845) 858-7707    8/1/1985   155.8      67   
219                                         NaN    4/9/1978   237.8      69   
229                 johndoe@email.com1234567890    1/1/1975   180.0      72   
230                                         NaN   9/23/1976   165.9      63   
234                                         NaN    4/7/1936   199.5      65   
237                 johndoe@email.com1234567890    1/1/1975   180.0      72   
242                                         NaN   2/25/1957   205.3      74   
244                 johndoe@email.com1234567890    1/1/1975   180.0      72   
249                                         NaN  10/30/1951   146.5      69   
251                 johndoe@email.com1234567890    1/1/1975   180.0      72   
257                                         NaN   5/17/1995   231.7      69   
264                                         NaN   11/3/1989   158.6      63   
269                                         NaN   10/9/1937   175.2      61   
277                 johndoe@email.com1234567890    1/1/1975   180.0      72   
278                                         NaN  12/16/1962   124.3      69   
282        304-438-2648SandraCTaylor@dayrep.com  10/23/1960   206.1      64   
286                                         NaN    4/1/1979 

In [15]:
patients.query('surname == "Jakobsen"')

patient_id assigned_sex given_name   surname             address  \
24           25         male      Jakob  Jakobsen   648 Old Dear Lane   
29           30         male       Jake  Jakobsen   648 Old Dear Lane   
432         433       female      Karen  Jakobsen  1690 Fannie Street   

            city     state  zip_code        country  \
24   Port Jervis  New York   12771.0  United States   
29   Port Jervis  New York   12771.0  United States   
432      Houston        TX   77020.0  United States   

                                        contact   birthdate  weight  height  \
24   JakobCJakobsen@einrot.com+1 (845) 858-7707    8/1/1985   155.8      67   
29   JakobCJakobsen@einrot.com+1 (845) 858-7707    8/1/1985   155.8      67   
432  KarenJakobsen@jourrapide.com1 979 203 0438  11/25/1962   185.2      67   

      bmi  
24   24.4  
29   24.4  
432  29.0

- value count for surname 'Doe' is 6
- Multiple records for Jakobsen, Taylor, Gersten

In [16]:
patients.contact.sample(5)

182    201-739-3931KaodilinakachukwuNnonso@gustr.com
142            516-740-5280FinleyChandler@dayrep.com
326              518-379-0603SuomaKoivunen@gustr.com
39              508-454-2027GanimeteScancar@cuvox.de
107              240-322-1398RoccoChristie@rhyta.com
Name: contact, dtype: object

- tidiness issue : there are 2 variables in contact column : e-mail and phone number

In [17]:
treatments.head()

given_name     surname    auralin    novodra  hba1c_start  hba1c_end  \
0   veronika    jindrová  41u - 48u          -         7.63       7.20   
1     elliot  richardson          -  40u - 45u         7.56       7.09   
2   yukitaka    takenaka          -  39u - 36u         7.68       7.25   
3       skye  gormanston  33u - 36u          -         7.97       7.62   
4     alissa      montez          -  33u - 29u         7.78       7.46   

   hba1c_change  
0           NaN  
1          0.97  
2           NaN  
3          0.35  
4          0.32

In [18]:
treatments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 280 entries, 0 to 279
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    280 non-null    object 
 1   surname       280 non-null    object 
 2   auralin       280 non-null    object 
 3   novodra       280 non-null    object 
 4   hba1c_start   280 non-null    float64
 5   hba1c_end     280 non-null    float64
 6   hba1c_change  171 non-null    float64
dtypes: float64(3), object(4)
memory usage: 15.4+ KB


- Erroneous datatype ( auralin, novodra)

In [19]:
treatments.describe()

hba1c_start   hba1c_end  hba1c_change
count   280.000000  280.000000    171.000000
mean      7.985929    7.589286      0.546023
std       0.568638    0.569672      0.279555
min       7.500000    7.010000      0.200000
25%       7.660000    7.270000      0.340000
50%       7.800000    7.420000      0.380000
75%       7.970000    7.570000      0.920000
max       9.950000    9.580000      0.990000

- a hba1c_change value about 0.4 is a success, however maximum value of hba1c_change is : 0.99 and 3rd quartile value is : 0.92 that suggest a massive skew and a big change wish is just implausible.

In [20]:

treatments.query('hba1c_change == 0.99').head(1)

given_name      surname auralin    novodra  hba1c_start  hba1c_end  \
32      laura  ehrlichmann       -  43u - 40u         7.95       7.46   

    hba1c_change  
32          0.99

In [21]:
hba1c_start = 7.95
hba1c_end = 7.46
hba1c_change = hba1c_start - hba1c_end
hba1c_change

0.4900000000000002

- hba1c_change is 0.49 instead of 0.99 it is calculated wrong. this can be a data entry error. (4s mistaken as 9s)

In [22]:
treatments.isnull().sum()

given_name        0
surname           0
auralin           0
novodra           0
hba1c_start       0
hba1c_end         0
hba1c_change    109
dtype: int64

- Missing values in hba1c_change

In [23]:

treatments[['auralin', 'novodra']].head()

auralin    novodra
0  41u - 48u          -
1          -  40u - 45u
2          -  39u - 36u
3  33u - 36u          -
4          -  33u - 29u

- The letter 'u' is starting on ending doses for Auralin and Novodra

In [24]:
patients[['given_name' , 'surname']].head()

given_name  surname
0        Zoe  Wellish
1     Pamela     Hill
2        Jae   Debord
3       Liêm     Phan
4        Tim  Neudorf

In [25]:
treatments[['given_name' , 'surname']].head()

given_name     surname
0   veronika    jindrová
1     elliot  richardson
2   yukitaka    takenaka
3       skye  gormanston
4     alissa      montez

- given name and surname in treatments table are all lowercase, but in the patients table they are not : this can be an issue if we want to join the two table using given_name and surname as a key

In [26]:
treatments.shape[0]

280

- missing records (280 instead of 350)

In [27]:
patients.contact.sample(5)

46          724-419-3583JavierMoquin@rhyta.com
499         928-284-4492RumanBisliev@gustr.com
232      KyoukoOno@superrito.com1 858 587 9945
69     JosephMDay@teleworm.us+1 (773) 615-9328
311     EricaMacDonald@teleworm.us601-389-7682
Name: contact, dtype: object

- there are different representation for telephone number

In [28]:
treatments.auralin.isnull().sum()

np.int64(0)

In [29]:
treatments.novodra.isnull().sum()


np.int64(0)

- the two lines above suggest that there are zero null entries for both auralin and novodra columns. However there are some entries with dashes in both columns wish should be nulls

In [30]:
treatments.sample(5)


given_name  surname    auralin    novodra  hba1c_start  hba1c_end  \
252      ellen    luman          -  40u - 39u         9.27       8.77   
8        saber   ménard          -  54u - 54u         8.08       7.70   
248    brianna  lansell  39u - 47u          -         9.48       9.06   
65   elisabeth   dimmen          -  41u - 36u         7.60       7.28   
177     samuel     blix  48u - 55u          -         7.97       7.56   

     hba1c_change  
252          0.50  
8             NaN  
248           NaN  
65           0.32  
177           NaN

- tidiness issues : for auralin and novodra columns there are 3 variables: - treatment : auralin or novodra - starting dose - ending dose

In [31]:
adverse_reactions.head()

given_name     surname           adverse_reaction
0      berta  napolitani  injection site discomfort
1       lena        baer               hypoglycemia
2     joseph         day               hypoglycemia
3     flavia  fiorentino                      cough
4    manouck     wubbels          throat irritation

In [32]:

patients[['given_name' , 'surname']].head()

given_name  surname
0        Zoe  Wellish
1     Pamela     Hill
2        Jae   Debord
3       Liêm     Phan
4        Tim  Neudorf

In [33]:

adverse_reactions[['given_name' , 'surname']].head()

given_name     surname
0      berta  napolitani
1       lena        baer
2     joseph         day
3     flavia  fiorentino
4    manouck     wubbels

- given name and surname in adverse reactions table are all lowercase, but in the patients table they are not : this can be an issue if we want to join the two table using given_name and surname as a key
- adverse_reaction columns of adverse_reactions table belongs to the treatments table beacause an adverse reaction of a treatment is naturally connected to this treatment : we should join treatments table and adverse reactions table

In [34]:
# checking duplicated columns in three tables 
all_columns = pd.Series(list(patients) + list(treatments) + list(adverse_reactions))
all_columns[all_columns.duplicated()]

14    given_name
15       surname
21    given_name
22       surname
dtype: object

- we should remove given_name and surnme from the treatment table and keep patients_id as primary key to ensure joining tables

## Quality issues summary

`patients` table
- Zip code is a float not a string
- Zip code has four digits sometimes
- Tim Neudorf height is 27 in instead of 72 in
- Full state names sometimes, abbreviations other times
- Dsvid Gustafsson
- Missing demographic information (address - contact columns) (can't clean)
- Erroneous datatypes (assigned sex, state, zip_code, and birthdate columns)
- Multiple phone number formats
- Default John Doe data
- Multiple records for Jakobsen, Gersten, Taylor
- kgs instead of lbs for Zaitseva weight

`treatments` table
- Missing HbA1c changes
- The letter 'u' in starting and ending doses for Auralin and Novodra
- Lowercase given names and surnames
- Missing records (280 instead of 350)
- Erroneous datatypes (auralin and novodra columns)
- Inaccurate HbA1c changes (leading 4s mistaken as 9s)
- Nulls represented as dashes (-) in auralin and novodra columns

`adverse_reactions` table
- Lowercase given names and surnames

## Tidiness issues summary

Three rules to check tidiness

1. Each variable forms a column
2. Each observation forms a row
3. Each type of observational unit forms a table



- contact column in patients table should be split into phone number and e-mail columns
- three variables in two columns in treatments table (treatment, start_dose, end_dose)
- adverse_reaction columns of adverse_reactions table belongs to the treatments table
- given_name and surname columns are duplicated in patients and treatments table

## Cleaning Data

Cleaning data is the third and final step in the data wrangling process. This is where the quality and tidiness issues identified in the assess step are remedied. It can be done manually in spreadsheet programs or text editors, but data cleaning is often best done using code and in three steps.

- Define how to clean issue in words,
- Code : convert these words to code,
- Test to make sure if that code worked.

The very first thing to do before any cleaning occurs is to make a copy of each piece of data

In [36]:

# Making a copy of each dataset
patients_clean = patients.copy()
treatments_clean = treatments.copy()
adverse_reactions_clean = adverse_reactions.copy()

It's usually best to find missing data first if you can. In other words, completeness issues, which is a part of data quality. For our clinical trial dataset here, we identified three completeness issues.

Two in the treatments table:

- the missing HbA1c changes, and
- the missing records having 280 instead of 350

Then one missing data issue in the patients table:

- the missing demographic information (address, city, state, zip code, country, and contact)

We just can't address this third issue : the missing demographic information. We really have no way of getting this information from these people, because We can't contact them.
We can clean the other two issues though, the two in the treatments table: Missing HbA1c changes and the missing records (280 instead of 350)

### treatments: Missing records (280 instead of 350)

The missing `treatments` records are stored in a file named `treatments_cut.csv`. Let's read this file

In [37]:
treatments_cut = pd.read_csv('treatments_cut.csv')

In [38]:
treatments_cut.head()

given_name    surname    auralin    novodra  hba1c_start  hba1c_end  \
0      jožka  resanovič  22u - 30u          -         7.56       7.22   
1  inunnguaq   heilmann  57u - 67u          -         7.85       7.45   
2      alwin   svensson  36u - 39u          -         7.78       7.34   
3        thể      lương          -  61u - 64u         7.64       7.22   
4     amanda    ribeiro  36u - 44u          -         7.85       7.47   

   hba1c_change  
0          0.34  
1           NaN  
2           NaN  
3          0.92  
4          0.38

In [40]:
treatments_cut.shape

(70, 7)

### Define

In [44]:
frames = [treatments_clean, treatments_cut]
treatments_clean = pd.concat(frames, ignore_index=True)

In [45]:
treatments_clean.shape

(350, 7)

### `treatments`: Missing HbA1c changes and Inaccurate HbA1c changes (leading 4s mistaken as 9s)

Note: the "Inaccurate HbA1c changes (leading 4s mistaken as 9s)" observation, which is an accuracy issue and not a completeness issue, is included in this header because it is also fixed by the cleaning operation that fixes the missing "Missing HbA1c changes" observation. Multiple observations in one Define, Code, and Test header occurs multiple times in this notebook.

### Define

Recalculate the `hba1c_change` column: `hba1c_start` minus `hba1c_end`.

### Code

In [46]:
treatments_clean.hba1c_change = (treatments_clean.hba1c_start - treatments_clean.hba1c_end)

### Test

In [47]:
treatments_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 350 entries, 0 to 349
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   given_name    350 non-null    object 
 1   surname       350 non-null    object 
 2   auralin       350 non-null    object 
 3   novodra       350 non-null    object 
 4   hba1c_start   350 non-null    float64
 5   hba1c_end     350 non-null    float64
 6   hba1c_change  350 non-null    float64
dtypes: float64(3), object(4)
memory usage: 19.3+ KB


In [48]:
treatments_clean.hba1c_change.describe()

count    350.000000
mean       0.396286
std        0.059283
min        0.200000
25%        0.350000
50%        0.390000
75%        0.440000
max        0.540000
Name: hba1c_change, dtype: float64

### Tidiness Data

After addressing missing data first, cleaning for tidiness is usually the next logical step. For this oral insulin clinical trial dataset, we detected and documented four sections of the dataset that weren't tidy:

- in the patients table, there are two variables in one column (contact column) : phone number and email
- in the treatment's table, there are three variables in two columns (treatment, start dose and end dose in the auralin and novodra columns).
- the adverse_reactions table shouldn't exist ( the adverse reaction column should be part of the treatments table)
- the given_name and surname columns are duplicated in the treatments and in adverse_reactions table when it should only be in the patients table.


###  Contact column in `patients` table contains two variables: phone number and email

### Define


Extract the phone number and email variables from the contact column using regular expressions and pandas' str.extract method. Drop the contact column when done.

### Code 

In [49]:

patients_clean['phone_number'] = patients_clean.contact.str.extract('((?:\+\d{1,2}\s)?\(?\d{3}\)?[\s.-]?\d{3}[\s.-]?\d{4})', expand=True)

# [a-zA-Z] to signify emails in this dataset all start and end with letters
patients_clean['email'] = patients_clean.contact.str.extract('([a-zA-Z][a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+[a-zA-Z])', expand=True)

# Note: axis=1 denotes that we are referring to a column, not a row
patients_clean = patients_clean.drop('contact', axis=1)

<>:1: SyntaxWarning: invalid escape sequence '\+'
<>:4: SyntaxWarning: invalid escape sequence '\.'
<>:1: SyntaxWarning: invalid escape sequence '\+'
<>:4: SyntaxWarning: invalid escape sequence '\.'
C:\Users\GT\AppData\Local\Temp\ipykernel_11672\3894566654.py:1: SyntaxWarning: invalid escape sequence '\+'
  patients_clean['phone_number'] = patients_clean.contact.str.extract('((?:\+\d{1,2}\s)?\(?\d{3}\)?[\s.-]?\d{3}[\s.-]?\d{4})', expand=True)
C:\Users\GT\AppData\Local\Temp\ipykernel_11672\3894566654.py:4: SyntaxWarning: invalid escape sequence '\.'
  patients_clean['email'] = patients_clean.contact.str.extract('([a-zA-Z][a-zA-Z0-9_.+-]+@[a-zA-Z0-9-]+\.[a-zA-Z0-9-.]+[a-zA-Z])', expand=True)


### Test

In [50]:
# Confirm contact column is gone
list(patients_clean)

['patient_id',
 'assigned_sex',
 'given_name',
 'surname',
 'address',
 'city',
 'state',
 'zip_code',
 'country',
 'birthdate',
 'weight',
 'height',
 'bmi',
 'phone_number',
 'email']

In [52]:
patients_clean.phone_number.sample(5)

244           1234567890
207         989-936-4563
330    +1 (843) 368-5129
264                  NaN
455         215-321-9611
Name: phone_number, dtype: object

In [53]:
patients_clean.email.sample(5)

421        ArminaSauve@dayrep.com
304    SignyGrimsdottir@rhyta.com
114    YunadiBarsukov@teleworm.us
35      KamilaPecinova@dayrep.com
48          AnnikaVaara@rhyta.com
Name: email, dtype: object

In [54]:
patients_clean.email.sort_values().head()

404               AaliyahRice@dayrep.com
11          Abdul-NurMummarIsa@rhyta.com
332                AbelEfrem@fleckens.hu
258              AbelYonatan@teleworm.us
305    AddolorataLombardi@jourrapide.com
Name: email, dtype: object


### Three variables in two columns in `treatments` table (treatment, start dose and end dose)
### Define
Melt the auralin and novodra columns to a treatment and a dose column (dose will still contain both start and end dose at this point). Then split the dose column on ' - ' to obtain start_dose and end_dose columns. Drop the intermediate dose column.

### Code

In [56]:
treatments_clean = pd.melt(treatments_clean, id_vars=['given_name', 'surname', 'hba1c_start', 'hba1c_end', 'hba1c_change'], var_name='treatment', value_name='dose')

In [60]:
treatments_clean = treatments_clean[treatments_clean['dose'] != '-']

In [62]:
treatments_clean[['start', 'end']] = treatments_clean['dose'].str.split('-', expand=True)

C:\Users\GT\AppData\Local\Temp\ipykernel_11672\1467536827.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  treatments_clean[['start', 'end']] = treatments_clean['dose'].str.split('-', expand=True)
C:\Users\GT\AppData\Local\Temp\ipykernel_11672\1467536827.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  treatments_clean[['start', 'end']] = treatments_clean['dose'].str.split('-', expand=True)


In [63]:
treatments_clean = treatments_clean.drop('dose', axis=1)

In [64]:
treatments_clean.head()

given_name     surname  hba1c_start  hba1c_end  hba1c_change treatment  \
0   veronika    jindrová         7.63       7.20          0.43   auralin   
3       skye  gormanston         7.97       7.62          0.35   auralin   
6     sophia      haugen         7.65       7.27          0.38   auralin   
7      eddie      archer         7.89       7.55          0.34   auralin   
9       asia     woźniak         7.76       7.37          0.39   auralin   

  start   end  
0  41u    48u  
3  33u    36u  
6  37u    42u  
7  31u    38u  
9  30u    36u

### Adverse reaction should be part of the `treatments` table
### Define
Merge the adverse_reaction column to the `treatments` table, joining on given name and surname.

### Code

In [65]:

treatments_clean = pd.merge(treatments_clean, adverse_reactions_clean, on=['given_name', 'surname'], how='left')

In [81]:
treatments_clean.sample(10)

given_name         surname  hba1c_start  hba1c_end  hba1c_change  \
0     veronika        jindrová         7.63       7.20          0.43   
192     cannan         cabrera         7.94       7.58          0.36   
61         rut  halldórsdóttir         9.01       8.63          0.38   
127       alex        crawford         7.69       7.30          0.39   
76       helen           luwam         7.66       7.19          0.47   
130       mika      martinsson         7.50       7.17          0.33   
249      ivona          jakšić         7.98       7.54          0.44   
97       minea        lindgren         9.45       8.94          0.51   
333      sofia       hermansen         8.90       8.57          0.33   
98       peter        pospíšil         7.76       7.34          0.42   

    treatment start   end           adverse_reaction  
0     auralin  41u    48u                        NaN  
192   novodra  40u    31u                        NaN  
61    auralin  31u    40u                        NaN  
127   auralin  51u    62u               hypoglycemia  
76    auralin  32u    42u                        NaN  
130   auralin  34u    43u                        NaN  
249   novodra  41u    41u                        NaN  
97    auralin  38u    45u                        NaN  
333   novodra  34u    34u  injection site discomfort  
98    auralin  51u    60u                        NaN

### Given name and surname columns in `patients` table duplicated in `treatments` and `adverse_reactions` tables and Lowercase given names and surnames

### Define
- Adverse reactions table is no longer needed.

- Isolate the patient ID and names in the patients table, then convert these names to lower case to join with treatments.
Drop the given name and surname columns in the treatments table (so these being lowercase isn't an issue anymore).
### Code

In [83]:
name_id = patients_clean[['patient_id', 'given_name', 'surname']]
name_id.given_name = name_id.given_name.str.lower()
name_id.surname = name_id.surname.str.lower()

treatments_clean = pd.merge(treatments_clean, name_id, on=['given_name', 'surname'])

treatments_clean = treatments_clean.drop(['given_name', 'surname'], axis=1)

C:\Users\GT\AppData\Local\Temp\ipykernel_11672\599544455.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  name_id.given_name = name_id.given_name.str.lower()
C:\Users\GT\AppData\Local\Temp\ipykernel_11672\599544455.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  name_id.surname = name_id.surname.str.lower()


### Test

In [84]:
treatments_clean.head()

hba1c_start  hba1c_end  hba1c_change treatment start   end  \
0         7.63       7.20          0.43   auralin  41u    48u   
1         7.97       7.62          0.35   auralin  33u    36u   
2         7.65       7.27          0.38   auralin  37u    42u   
3         7.89       7.55          0.34   auralin  31u    38u   
4         7.76       7.37          0.39   auralin  30u    36u   

  adverse_reaction  patient_id  
0              NaN         225  
1              NaN         242  
2              NaN         345  
3              NaN         276  
4              NaN          15

In [85]:

# Patient ID should be the only duplicate column
all_columns = pd.Series(list(patients_clean) + list(treatments_clean))
all_columns[all_columns.duplicated()]

22    patient_id
dtype: object

### Quality
Once the missing data and tidiness issues are cleaned, cleaning the remaining data quality issues, i.e.the ones outside of the completeness issues is all that remains.

There were 19 data quality issues in this dataset (16 outside of the missing data issues). Next we'll use a variety of common pandas functions to clean them all.

### Zip code is a float not a string and Zip code has four digits sometimes
### Define
- Convert the zip code column's data type from a float to a string using (astype)
- Remove the .0 using string slicing and pad four digit zip codes witha leading 0
### Code

In [86]:
patients_clean.zip_code = patients_clean.zip_code.astype(str).str[:-2].str.pad(5, fillchar='0')

In [112]:
patients_clean.zip_code.sample()

462    90007
Name: zip_code, dtype: object

### Tim Neudorf height is 27 in instead of 72 in
### Define
Replace height for rows in the patients table that have a height of 27 in (there is only one) with 72 in.
### Code

In [113]:
patients_clean['height'].replace(27, 72, inplace=True)

C:\Users\GT\AppData\Local\Temp\ipykernel_11672\1100413820.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  patients_clean['height'].replace(27, 72, inplace=True)


### Test

In [114]:
patients_clean.query('height == "27"')

Empty DataFrame
Columns: [patient_id, assigned_sex, given_name, surname, address, city, state, zip_code, country, birthdate, weight, height, bmi, phone_number, email]
Index: []

In [115]:

# Confirm the replacement worked
patients_clean[patients_clean.surname == 'Neudorf'].height

4    72
Name: height, dtype: int64

### Full state names sometimes, abbreviations other times
### Define
Apply a function that converts full state name to state abbreviation for California, New York, Illinois, Florida, and Nebraska.
### Code

In [118]:
# Mapping from full state name to abbreviation
state_abbrev = {'California': 'CA',
                'New York': 'NY',
                'Illinois': 'IL',
                'Florida': 'FL',
                'Nebraska': 'NE'}

# Function to apply
def abbreviate_state(patient):
    if patient['state'] in state_abbrev.keys():
        abbrev = state_abbrev[patient['state']]
        return abbrev
    else:
        return patient['state']
    
patients_clean['state'] = patients_clean.apply(abbreviate_state, axis=1)

### Test

In [119]:
patients_clean.state.value_counts()

state
CA    60
NY    47
TX    32
IL    24
FL    22
MA    22
PA    18
GA    15
OH    14
LA    13
MI    13
OK    13
NJ    12
VA    11
MS    10
WI    10
TN     9
IN     9
AL     9
MN     9
KY     8
NC     8
WA     8
MO     7
NV     6
ID     6
KS     6
NE     6
IA     5
SC     5
CT     5
AR     4
RI     4
ME     4
CO     4
ND     4
AZ     4
OR     3
MD     3
SD     3
DE     3
WV     3
MT     2
VT     2
DC     2
NM     1
WY     1
AK     1
NH     1
Name: count, dtype: int64


### Dsvid Gustafsson
### Define
Replace given name for rows in the `patients` table that have a given name of 'Dsvid' with 'David'.

### Code

In [120]:
patients_clean.given_name.replace('Dsvid', 'David', inplace=True)

### Test

In [121]:
patients_clean.query('given_name == "Dsvid"')

Empty DataFrame
Columns: [patient_id, assigned_sex, given_name, surname, address, city, state, zip_code, country, birthdate, weight, height, bmi, phone_number, email]
Index: []

In [122]:
patients_clean.query('surname == "Gustafsson"').given_name

8    David
Name: given_name, dtype: object

### Erroneous datatypes (assigned sex, state, zip_code, and birthdate columns) and Erroneous datatypes (auralin and novodra columns) and The letter 'u' in starting and ending doses for Auralin and Novodra
### Define
Convert assigned sex and state to categorical data types. Zip code data type was already addressed above. Convert birthdate to datetime data type. Strip the letter 'u' in start dose and end dose and convert those columns to data type integer.

### Code

In [133]:
patients_clean.assigned_sex = patients_clean.assigned_sex.astype('category')
patients_clean.state = patients_clean.state.astype('category')

patients_clean['birthdate'] = pd.to_datetime(patients_clean['birthdate'])

treatments_clean['start'] = treatments_clean['start'].str.replace(r'[^\d]', '', regex=True).astype(int)
treatments_clean['end'] = treatments_clean['end'].str.replace(r'[^\d]', '', regex=True).astype(int)

### Test

In [135]:
patients_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 503 entries, 0 to 502
Data columns (total 15 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   patient_id    503 non-null    int64         
 1   assigned_sex  503 non-null    category      
 2   given_name    503 non-null    object        
 3   surname       503 non-null    object        
 4   address       491 non-null    object        
 5   city          491 non-null    object        
 6   state         491 non-null    category      
 7   zip_code      503 non-null    object        
 8   country       491 non-null    object        
 9   birthdate     503 non-null    datetime64[ns]
 10  weight        503 non-null    float64       
 11  height        503 non-null    int64         
 12  bmi           503 non-null    float64       
 13  phone_number  491 non-null    object        
 14  email         491 non-null    object        
dtypes: category(2), datetime64[ns](1), float

In [136]:
treatments_clean[['start', 'end']].head()

start  end
0     41   48
1     33   36
2     37   42
3     31   38
4     30   36

In [137]:

treatments_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 349 entries, 0 to 348
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   hba1c_start       349 non-null    float64
 1   hba1c_end         349 non-null    float64
 2   hba1c_change      349 non-null    float64
 3   treatment         349 non-null    object 
 4   start             349 non-null    int64  
 5   end               349 non-null    int64  
 6   adverse_reaction  35 non-null     object 
 7   patient_id        349 non-null    int64  
dtypes: float64(3), int64(3), object(2)
memory usage: 21.9+ KB


### Multiple phone number formats
### Define
- Strip all " ", "-", "(", ")", and "+" and store each number without any formatting. Pad the phone number with a 1 if the length of the number is 10 digits (we want country code). 

### code

In [140]:
patients_clean.phone_number = patients_clean.phone_number.str.replace(r'\D+', '', regex=True).str.pad(11, fillchar='1')


### Test

In [141]:
patients_clean.phone_number.head()


0    19517199170
1    12175693204
2    14023636804
3    17326368246
4    13345157487
Name: phone_number, dtype: object

### Default John Doe data

### Define
Remove the non-recoverable John Doe records from the `patients` table.
### Code


In [142]:

patients_clean = patients_clean[patients_clean.surname != 'Doe']

### Test

In [143]:
patients_clean[patients_clean.surname == 'Doe']

Empty DataFrame
Columns: [patient_id, assigned_sex, given_name, surname, address, city, state, zip_code, country, birthdate, weight, height, bmi, phone_number, email]
Index: []


### Multiple records for Jakobsen, Gersten, Taylor
### Define
- Remove the Jake Jakobsen, Pat Gersten, and Sandy Taylor rows from the patients table. These are the nicknames, which happen to also not be in the treatments table (removing the wrong name would create a consistency issue between the patients and treatments table). These are all the second occurrence of the duplicate. These are also the only occurences of non-null duplicate addresses.
### Code

In [144]:
patients_clean = patients_clean[~((patients_clean.address.duplicated()) & patients_clean.address.notnull())]


### Test

In [145]:
patients_clean[patients_clean.surname == 'Jakobsen']


patient_id assigned_sex given_name   surname             address  \
24           25         male      Jakob  Jakobsen   648 Old Dear Lane   
432         433       female      Karen  Jakobsen  1690 Fannie Street   

            city state zip_code        country  birthdate  weight  height  \
24   Port Jervis    NY    12771  United States 1985-08-01   155.8      67   
432      Houston    TX    77020  United States 1962-11-25   185.2      67   

      bmi phone_number                         email  
24   24.4  18458587707     JakobCJakobsen@einrot.com  
432  29.0  19792030438  KarenJakobsen@jourrapide.com

In [146]:

patients_clean[patients_clean.surname == 'Gersten']

patient_id assigned_sex given_name  surname            address  city  \
97          98         male    Patrick  Gersten  2778 North Avenue  Burr   

   state zip_code        country  birthdate  weight  height   bmi  \
97    NE    68324  United States 1954-05-03   138.2      71  19.3   

   phone_number                     email  
97  14028484923  PatrickGersten@rhyta.com

In [147]:

patients_clean[patients_clean.surname == 'Taylor']

patient_id assigned_sex given_name surname             address      city  \
131         132       female     Sandra  Taylor  2476 Fulton Street  Rainelle   
426         427         male    Rogelio  Taylor  4064 Marigold Lane     Miami   

    state zip_code        country  birthdate  weight  height   bmi  \
131    WV    25962  United States 1960-10-23   206.1      64  35.4   
426    FL    33179  United States 1992-09-02   186.6      69  27.6   

    phone_number                       email  
131  13044382648    SandraCTaylor@dayrep.com  
426  13054346299  RogelioJTaylor@teleworm.us

### kgs instead of lbs for Zaitseva weight
### Define
Use advanced indexing to isolate the row where the surname is Zaitseva and convert the entry in its weight field from kg to lbs.

### Code

In [148]:

weight_kg = patients_clean.weight.min()
mask = patients_clean.surname == 'Zaitseva'
column_name = 'weight'
patients_clean.loc[mask, column_name] = weight_kg * 2.20462

### Test

In [149]:
patients_clean.weight.min()

np.float64(102.1)